_posts/2016-04-12-ecto-model-schema-extension.md

```diff
@@ -0,0 +1,124 @@
+
+### How
+
+There are three parts needed at minimum to create & use an extension effectively:
+
+- Library Code
+- Service Code
+- Consumer Code
+
+An extension and its use with Nectar can be viewed as Producer / Consumer relationship bound by a communication protocol.
+
+**Extension** which want to add a virtual field, say special, to Nectar Product Model Schema is a **Producer (Service Code)**.
+
+**Nectar Model** is a **Consumer (Consumer Code)** allowing the schema changes through a **communication protocol (Library Code)**
+
+Let's begin the journey of incremental changes to bring consumer, service and library code into existence starting from a simple use-case of adding a virtual boolean field, say special to Nectar Product.
+
+>
+Note: Please refer [Intro to Metaprogramming]() for more information on Metaprogramming in Elixir
+
+1.  A straightforward way to add a virtual field, say special, to Nectar Product would be to add it directly in Nectar.Product Schema definition, but it requires change in Nectar source. Let's move to next step for the workaround to avoid modification to Nectar.Product
+
+    <script src="https://gist.github.com/pikender/f58b2208ae8951c7b13214bf320e8ec1/2faba2e7a14bb77cceec769ef676fd439244878d.js"></script>
+
+    <script src="https://gist.github.com/pikender/bf89a77d2ed7c684dd0258d88e777cc0.js"></script>
+
+1.  We can add a function to Nectar Model Schema to which other extensions can delegate the responsibility of schema changes, see full version [here](https://gist.github.com/pikender/f58b2208ae8951c7b13214bf320e8ec1/5f4ada57be942f8dce713cf4c6c0d6761a7632a0). See Nectar.ExtendProduct example below on how to use it.
+
+    <script src="https://gist.github.com/pikender/cb43c04937fbb95b289bfa43d8dfab08/867502fb2218c41b6495bf318fab527a8a185193.js"></script>
+
+    <script src="https://gist.github.com/pikender/bf89a77d2ed7c684dd0258d88e777cc0.js"></script>
+
+1.  Now, with delegation function `extensions` in place, we can work towards providing a way to register the schema changes, see full version [here](https://gist.github.com/pikender/f58b2208ae8951c7b13214bf320e8ec1/d2b931acb891d74014d2c5f6a1996f69c222e01c). Please check the usage of Module attributes for same below.
+
+    <script src="https://gist.github.com/pikender/cb43c04937fbb95b289bfa43d8dfab08/51619c29cf3a741c85b64e8e6e6ee254457393c5.js"></script>
+
+    <script src="https://gist.github.com/pikender/bf89a77d2ed7c684dd0258d88e777cc0.js"></script>
+
+1.  Earlier, Module.put_attribute need to be used multiple times to define multiple schema changes instead we wrapped it in an anonymous function to encapsulate the collection of schema changes through a simple and consistent interface, see full version [here](https://gist.github.com/pikender/f58b2208ae8951c7b13214bf320e8ec1/3312acebeb9edec66e61da2ad447f7b18d5a9c8e). There can be multiple extensions used for different functionality and hence multiple schema changes need to be registered and defined
```

Comment:
Earlier, Module.put_attribute was used multiple times to define multiple schema changes whereas now we wrapped it in an anonymous function to encapsulate the collection of schema changes through a simple and consistent interface, see full version [here]